core/src/main/scala/dogs/Memo.scala

```diff
@@ -0,0 +1,248 @@
+package dogs
+
+import scala.collection.{ generic, mutable, immutable, concurrent, JavaConverters }
+import scala.reflect.ClassTag
+
+/** A memo is a recipe for creating a memoized version of a function.
+  *
+  * The contract is that you pass in a function and get back a new version of that function that
+  * uses the specified mechanism to cache results. A `Memo` doesn't necessarily guarantee that a
+  * function will be called only once for a given input: each memo will have its own behavior with
+  * respect to this.
+  *
+  * This provides a separation of the way a function is memoized from the point
+  * where it happens. This both reduces boilerplate calls to functions like
+  * [[scala.collection.mutable.Map.getOrElseUpdate `getOrElseUpdate`]] and allows for the
+  * possibility of switching out implementations dynamically or with a single change.
+  *
+  * @example
+  * {{{
+  * import dogs.Memo
+  *
+  * // Memoize a function directly
+  * val fastValue = Memo.mutableHashMapMemo { s: String => ??? }
+  *
+  * // Abstract over memoization strategy
+  * val memoize = Memo.concurrentTrieMapMemo
+  * val getValue = memoize { a: Int => ??? }
+  *
+  * // Memoize an existing function
+  * def myFunction(s: String): Int = ???
+  * val myFastFunction = memoize(myFunction)
+  * }}}
+  *
+  * @tparam A the upper bound of the domain values that this `Memo` can store
+  * @tparam B the upper bound of the range values that this `Memo` can store
+  *
+  * @note Implementors should ensure that a `Memo` produces a new backing store each time
+  * [[Memo.apply apply]] is called so that it cannot produce incorrect results if two functions
+  * disagree about the output for a given input. The `Memo` is the recipe, not the data store.
+  */
+trait Memo[A, B] {
+  def apply[A1 <: A, B1 <: B](fn: A1 => B1): A1 => B1
```

Comment:
@stew I agree that the type variance is a bit counterintuitive, but the type bounds are boundaries based on the storage. The reason to have the trait is that it turns a `Memo` object into a strategy object independent of the actual function that's memoized. Perhaps this would be more obvious if it were named `MemoStrategy`. 

An illustrating example is the `arrayMemo[A]`, which stores data into a backing store of `Array[A]`, effectively `Int => A`. The domain parameter is intrinsically covariant: you can't put an `Any => Any` into an `arrayMemo`. Yes, a `Memo[Any, Any]` can memoize any function—but a `Memo[Int, Any]` cannot.

It's true that you could assign an `Any => String` as an `Int => String`, effectively allowing you to pass a `Any => String` parameter into an `arrayMemo`, but you still don't violate type safety because the output of the memo is an `Int => String`: It's constraining the domain of the _output_ function, not the input function.

It doesn't work to take the parameters off of the `apply` because you'd wind up with something like `arrayMemo`, which has a signature bound of `Int => Any`, producing a memoized function that always returns an `Any` even if you passed in an `Int => Int`. 

Scalaz's parameters are on its trait, but they do this by requiring you to effectively create a new `Memo` object for every function you want to memoize. You can't do something like the below with Scalaz unless you create a new abstraction of your own.

```
val memoize = if (config.useMemos) Memo.mutableHashMapMemo else Memo.nilMemo

val getScore: String => Int = memoize(…)
val fetchProfile: UserId => UserProfile = memoize(…)
```

There are basically two design goals that necessitate the type variance I've designed:
  * Separation of concerns: a `Memo` is a storage strategy that can be reused to memoize several functions that don't all have identical types.
    * (This requires that you reparameterize at the time of application.)
  *  Storage specialization: memo strategies like `arrayMemo` that don't allow you to assume that all `Memo` objects have type `Memo[Any, Any]`.

I would also point out that the tricky types here are really only an issue for the library implementer, and not so much for the caller. A user doesn't care about abstracting over strategy can just call `Memo.mutableHashMapMemo` at the application point without ever worrying about the storage-type boundaries.

Put another way, from a user's perspective, the extra type bounds allow for more advanced use cases, but do not require *anything* extra of somebody who wants to use it in way you'd use a Scalaz memo. A Scalaz-style system cannot support `Memo` reuse, but the system I've proposed can support Scalaz-style use.

*I can draft something with a couple of type aliases that might make the use cases clearer if that would help?* What do you think?